plutus-benchmark/flat/Main.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE BangPatterns      #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main where
+
+import           Control.DeepSeq              (force)
+import           Criterion.Main
+import           Criterion.Types
+import qualified Data.ByteString              as BS
+import           Data.Map.Strict              (Map, fromList, lookup)
+import           Data.Maybe                   (fromJust)
+import           Data.Text                    (Text, unpack)
+import           System.Mem                   (performMajorGC)
+
+import           Language.PlutusCore          (Name (..))
+import           Language.PlutusCore.Universe
+import           Language.UntypedPlutusCore
+
+import           Codec
+import           Dataset
+import           Report
+
+import           Prelude                      hiding (lookup)
+
+type Tm = Term Name DefaultUni ()
+
+serializedContracts :: Map (Text, Text) BS.ByteString
+serializedContracts = fromList $ do
+  (contractName, contract) <- contracts
+  (codecName, codec)       <- codecs
+  pure $ ((contractName, codecName), serialize codec contract)
+
+benchmarkSerialize ::
+     (Text, Tm)
+  -> (Text, Codec Tm)
+  -> Benchmark
+benchmarkSerialize (contractName, contract) (codecName, (Codec serialize' _)) =
+  bench (unpack $ " " <> contractName <> " - " <> codecName)
+        -- (nf (BS.length . serialize) contract) ~ original code
+        (nf serialize' contract)
+
+benchmarkDeserialize ::
+     Map (Text, Text) BS.ByteString
+  -> (Text, Tm)
+  -> (Text, Codec Tm)
+  -> Benchmark
+benchmarkDeserialize scs (contractName, _) (codecName, codec) =
+  let !serialized = fromJust $ lookup (contractName, codecName) scs
+  in  bench (unpack $ " " <> contractName <> " - " <> codecName)
+            -- (whnf ((contract ==) . (Codec.deserialize codec)) serialized) ~ original code
+            (whnf (Codec.deserialize codec) serialized)
+
+benchmarkContract ::
+     Map (Text, Text) BS.ByteString
+  -> (Text, Tm)
+  -> [Benchmark]
+benchmarkContract scs contract =
+  [ bgroup "serialization (time) " $
+       map (benchmarkSerialize contract) codecs
+   , bgroup "deserialization (time) " $
+       map (benchmarkDeserialize scs contract) codecs
+  ]
+
+benchmarkConfig :: Config
+benchmarkConfig = defaultConfig
+  { jsonFile   = Just "flat-report.json"
+  , reportFile = Just "flat-report.html"
+  }
+
+sizesReport :: FilePath
+sizesReport = "flat-sizes.txt"
+
+main :: IO ()
+main = do
+  let !scs = force serializedContracts
+      !contracts' = force contracts
+  performMajorGC
```

Comment:
Various mysterious things happen here and above: comments!

Reply:
I am not sure what is mysterious, but I have added a comment. Hopefully, that solves some of the mystery.

Reply:
Well: I look at that and I think "Radu certainly had some reason for adding these bang patterns, forces, and the GC call. But I don't know what they were, so I can't comment if this is sufficient, and I don't know if it's safe to touch this."

Reply:
I am actually not sure about the `performMajorGC` bit. Is taken from the original tests. I will remove it and see what happens.

Reply:
"I took it from the original benchmarks and I'm not sure whether it is needed" is itself a very useful comment!

Reply:
Removing `performMajorGC` does not seem to affect the tests. So I won't add it back.